### PR TITLE
Add updated `Modal` component and documentation

### PR DESCRIPTION
**Pull request by lyzadanger**
Depends on #825

Don't panic: More than half the diff is the pattern-library documentation.

This PR adds a `Modal` component that conforms to the current API and styling conventions for shared components. It combines the functionality of legacy `Dialog` and `Modal` components[^1].

image

This implementation should be considered a bridge implementation until we have a little time to 
focus on a full-featured `dialog` component.

It retains from `Dialog` and `Modal`, unchanged:

* initial focus-routing behavior
* `aria-describedby` behavior
* `useElementShouldClose` handling
* `Modal`s use a `Panel` layout still

What it does add is, I hope, more sane modal sizing and positioning configuration. There are some size defaults, and a `width` prop, and some styling guard rails to ensure that modal dialogs never get too big or stupid for their viewports. I've combined sizing and positioning use cases from LMS, the client and the previous implementations of this component. The API is simpler than previously.

The [pattern-library page](http://localhost:4001/feedback-modal) for this new component documents and exemplifies its aims. I suggest starting there...

[^1]: `Dialog` is not actually used anywhere in our applications directly. Legacy `Modal` wraps `Dialog`.

### CODE REVIEW

File: src/components/feedback/Modal.tsx
Comment (lyzadanger) [inline src/components/feedback/Modal.tsx:60]: This `useEffect` is copied from previous implementation of `Dialog`.

File: src/components/feedback/Modal.tsx
Comment (lyzadanger) [inline src/components/feedback/Modal.tsx:79]: This `useLayoutEffect` is copied from a previous implementation of `Dialog`

File: src/components/feedback/test/Modal-test.js
Comment (lyzadanger) [inline src/components/feedback/test/Modal-test.js:22]: This `describe` block is essentially copied from pre-existing `Dialog` tests

File: src/components/feedback/test/Modal-test.js
Comment (lyzadanger) [inline src/components/feedback/test/Modal-test.js:90]: These tests were copied, too.

File: src/components/feedback/test/Modal-test.js
Comment (lyzadanger) [inline src/components/feedback/test/Modal-test.js:112]: I wrapped these tests in a `describe` block, but they already existed on the previous `Dialog` implementation.

File: src/tailwind.preset.js
Comment (lyzadanger) [inline src/tailwind.preset.js:87]: This "screen" (media query) is used by the new `Modal`

File: src/components/feedback/Modal.tsx
Comment (acelaya) [inline src/components/feedback/Modal.tsx:28]: I see this prop handles three possible "states":

* A ref is provided, in which case the referenced component gets focused.
* The prop is not provided at all, in which case the modal itself gets focused.
* The `null` value is provided, in which case we opt out from focusing capabilities.

I think the last two can be easily confused in consumer code. If I see `<Modal initialFocus={null} ... />` I will not remember that means the focusing is disabled.

I don't see an easy solution though. Maybe accepting a constant value instead, which is more clear in consumer code? Something in the lines of `<Modal initialFocus="DISABLE_FOCUS" ... />`. It's a bit hacky, but maybe more descriptive.

Other ideas?

File: src/components/feedback/Modal.tsx
Comment (acelaya) [inline src/components/feedback/Modal.tsx:73]: Can I suggest not nesting `if`s? I find it easier to use early returns so that you can mentally discard conditions as soon as possible, as you "follow" the code. When nesting, you need to "drag" all of them.

```ts
if (initialFocus === null) {
  return;
}

const focusEl = initialFocus?.current as HTMLInputElement;
if (focusEl && !focusEl.disabled) {
    focusEl.focus();
} else {
  // ...
  modalRef.current?.focus();
}
```

**Review by acelaya — CHANGES_REQUESTED** [review]
Looks good in general.

I just added two minor comments, but they are very much personal preference. Feel free to answer to those and I'm open to approve if the general feeling is that they make no sense.

**Comment by robertknight** [comment]
The combination of single line separator that doesn't cover the full width of the content, and top-shadow for scrolled content, which does, can look a bit odd:

Top shadow and underline

File: src/components/feedback/Modal.tsx
Comment (robertknight) [reply src/components/feedback/Modal.tsx:28]: > The prop is not provided at all, in which case the modal itself gets focused.
> The null value is provided, in which case we opt out from focusing capabilities.

What uses do we have for these two separate cases? Could we just always use one of the two behaviors here?

File: src/components/feedback/Modal.tsx
Comment (robertknight) [inline src/components/feedback/Modal.tsx:63]: Hmm... the assumption that `focusEl` is an `HTMLInputElement` might not always be valid. For example it might be a textarea or button. I presume the reason we are doing this is that not all HTML elements have a `disabled` prop. We could make this clearer with something like:

```js
      const focusEl = initialFocus?.current as HTMLElement & { disabled?: boolean };
      if (focusEl && !focusEl.disabled) {
```

**Review by robertknight — COMMENTED** [review]
This generally looks good. I think the most important thing to resolve once we get this landed will be to properly implement focus trapping etc, so that our dialogs are following correct behavior for keyboard navigation.

File: src/components/feedback/Modal.tsx
Comment (lyzadanger) [reply src/components/feedback/Modal.tsx:28]: Good feedback, thanks!

File: src/components/feedback/Modal.tsx
Comment (lyzadanger) [reply src/components/feedback/Modal.tsx:73]: Sure! I kind of blindly cut-and-pasted this from a previous implementation so it's good to get fresh eyes on it. I agree. Guards are easier to reason about.

File: src/components/feedback/Modal.tsx
Comment (lyzadanger) [reply src/components/feedback/Modal.tsx:63]: > The combination of single line separator that doesn't cover the full width of the content, and top-shadow for scrolled content, which does, can look a bit odd:
> 
> Top shadow and underline

I tend to agree. However, this styling is actually owned by `Panel` so I want to hit this in a follow-up. Making a note.

**Comment by lyzadanger** [comment]
I've pushed a commit that adjusts `initialFocus` handling per feedback.

**Review by acelaya — APPROVED** [review]
Looks good!

**Pull request merged**
